### ISSUE TITLE
RFC: big picture / fullscreen mode

### ISSUE THREAD

**Issue by neuromancer** [tkashkin/GameHub#75]
A big picture (or just fullscreen) mode for launching games will be a really nice addition. Some important features are:

- [ ] enable this mode from startup
- [ ] allow offline mode
- [ ] start showing only installed games
- [ ] filter, navigate and launch games with gamepad or keyboard (mouse should be optional)

**Comment by tkashkin** [comment]
I have implemented very basic keyboard navigation in grid view.
Use `arrows` to navigate, `return` or `space` to launch game, `ctrl` or `menu` to show context menu.

Basic gamepad navigation is also supported, but may not work. I haven't found a way to emit keypresses reliably.

It requires `xdotool` to emit keypresses, obviously it shouldn't work on wayland (idk if anything works in wayland :slightly_smiling_face:).

Use `D-Pad` to navigate (no stick support yet), `A` to launch game, `X` to show context menu, `B` to hide context menu.
https://github.com/tkashkin/GameHub/blob/4c29f49e0b5e78ae381467e05e2c1f9c92faeff4/src/utils/Gamepad.vala

**Comment by neuromancer** [comment]
Oh, cool. I can test with my steam controller, but it will be nice to have someone else using a different one. I will test this next week since I'm away from my desktop computer.

**Comment by tkashkin** [comment]
Stick support is added in e3d214c. It may be buggy.

How should UI update when controller is connected/disconnected?
Any ideas for bindings (besides `dpad/LS -> arrows`, `A - return`, `B - esc`, `X - menu`)?

`hid_steam` maps Steam Controller grips and even trackpad touch as separate buttons, but unfortunately it doesn't seem to map left trackpad as separate axes.
It's possible to know if left/right pad was touched/released, but impossible to know where exactly it was touched :disappointed:.
It also maps left pad as a dpad if pressed (I don't like "requires press" trackpads mode).
Right pad is mapped as a right stick.

**Comment by neuromancer** [comment]
> Any ideas for bindings (besides dpad/LS -> arrows, A - return, B - esc, X - menu)?

Maybe `Y` could be used like the left click to open some menu for every game.

Also, to test with some additional configurations, you can use sc-controller to emulate a Xbox controller.

**Comment by tkashkin** [comment]
UI now switches to controller mode when gamepad is connected and back to normal mode when it's disconnected:

![image](https://user-images.githubusercontent.com/2003017/46922405-6d34d300-d011-11e8-8a80-09dc0f21f4ed.png)

**Comment by neuromancer** [comment]
It looks great!

You can also consider adding some sorting features, like [Playnite implemented](https://www.youtube.com/watch?v=P2cXQXbIuLc).

**Comment by MaoDzeDun** [comment]
Yes, thank you - you're building the unified gaming library I've dreamed of. Now make a Patreon already :D.

**Comment by neuromancer** [comment]
Maybe you can try [gitcoin](https://gitcoin.co/) ?

**Comment by MaoDzeDun** [comment]
@tkashkin I think you will be surprised. I for one would definitely would. Please, look into it. I think you are very talented and deserve the financial incentive.

**Comment by tkashkin** [comment]
@MaoDzeDun I've setup Patreon and Liberapay pages (links in readme: https://github.com/tkashkin/GameHub/tree/dev#donate).
However I can't be sure if it will work or if I will be able to withdraw donations.
You can test it if you have some spare money that is not guaranteed to get to me. :slightly_smiling_face:

**Comment by MaoDzeDun** [comment]
@tkashkin There - 2.50$ a month from me (3 with VAT :( ). Wish I could donate more but I'm a father of a baby living in Bulgaria (i.e. I'm poor :D). You might want to put them on top (if possible) though - 9 out of 10 people don't scroll to the page bottom. Also, as a recommendation per personal experience - building has been the only install option that worked for me on Linux Mint 19. Trying to install through PPA gives an unmet dependency error - libgranite5, which I haven't found a way to install... though I am probably dumb and should have searched more into the problem. Anyway, you'd probably get a higher adoption rate if you sort Ubuntu install - us Ubuntu-based people are simple folk :D.

**Comment by tkashkin** [comment]
@MaoDzeDun Thanks. I'll update when it will be actually possible to try to withdraw it to PayPal and then to somehow use or withdraw it to card from PayPal.

As for `libgranite5`, it's gtk extension library for elementary OS. It's interesting that you are able to run version you've built. Maybe you have `libgranite4` installed.
You can try to install `libgranite5` from [here](https://debian.pkgs.org/sid/debian-main-amd64/libgranite5_5.1.0-1_amd64.deb.html) or use AppImage versions from releases.
You can also add elementary PPA and install it from here, but it's not a good thing to mix distro-specific repositories.
I'll try to somehow add granite and other dependencies in my PPA.

**Comment by tkashkin** [comment]
@MaoDzeDun I have added `granite` to my PPA, you should now be able to install GameHub directly from PPA without having `libgranite5` installed from elsewhere.

I have also sent you a message on Patreon, what do you think I should do?

**Comment by MaoDzeDun** [comment]
@tkashkin Thank you - I will test it out tonight when I get home from work.
Just replied to you message on Patreon :).

**Comment by neuromancer** [comment]
It works! (so I updated the lists of task!).

**Comment by Hedronmx** [comment]
Can't open settings when controller is plugged in. If using steam controller, the controller overlay is always turned on. Or at least that's the case for me...

**Comment by Hedronmx** [comment]
Switch to controller mode only after any button is pressed and
Show settings button in all modes (and probably add some shortcut (select button?))

Like that :)

**Comment by tkashkin** [comment]
In e5b5298 settings button is always visible and has controller shortcut (`select`).
Also triggers now emit `Tab` and `Shift+Tab` to control focus and make navigation easier in some cases.

**Comment by neuromancer** [comment]
Last revision does not switch to controller mode. My steam controller can navigate and launch games (but it fails to change tabs). Maybe ArchLinux is directly using the steam controller as a keyboard/mouse?

**Comment by tkashkin** [comment]
@neuromancer 

- What version are you using?
- If you have built it yourself, do you have `libmanette` installed and used during build?
- Is there anything gamepad-related in log?
- There may be some bugs which are already fixed.

**Comment by neuromancer** [comment]
Btw, X and Y seem to be wrong: Y opens the contextual menu per game and X opens the filtering menu. Is this expected?

**Comment by tkashkin** [comment]
Interesting. It was correct for me with Steam Controller with both `hid_steam` kernel module and with `sc-controller`.
It should print pressed keys and its values in log if app is started with `--debug` argument.

**Comment by neuromancer** [comment]
Interestingly enough, this happens if you use `sc-controller` in xbox mode. Otherwise, it works fine.

**Comment by tkashkin** [comment]
Yes, I can confirm. `X` and `Y` seems to be swapped between `hid_steam` and `sc_controller`.

###### Steam Controller (Wireless) with `hid_steam`:
```
'Wireless Steam Controller' connected

Button pressed: A (A) [304]
Button pressed: B (B) [305]
Button pressed: X (X) [307]
Button pressed: Y (Y) [308]

Button pressed: Select (Select) [314]
Button pressed: Start (Start) [315]
Button pressed: Home (Home) [316]

Button pressed: LT (Left Trigger) [312]
Button pressed: RT (Right Trigger) [313]

Button pressed: LB (Left Bumper) [310] 
Button pressed: RB (Right Bumper) [311]

Button pressed: LG (Left Grip) [336]
Button pressed: RG (Right Grip) [337]

Button pressed: LS (Left Stick) [317]
Button pressed: RS (Right Stick) [318]

Button pressed: L (Left Trackpad Touch) [289]
Button pressed: R (Right Trackpad Touch) [290]

Button pressed: Up (D-Pad Up) [544]
Button pressed: Down (D-Pad Down) [545]
Button pressed: Left (D-Pad Left) [546]
Button pressed: Right (D-Pad Right) [547]
```

###### Steam Controller (Wireless) with `sc-controller`:
```
'Microsoft X-Box 360 pad' connected

Button pressed: A (A) [304] // A
Button pressed: B (B) [305] // B
Button pressed: Y (Y) [308] // X
Button pressed: X (X) [307] // Y

Button pressed: Select (Select) [314]
Button pressed: Start (Start) [315]

Button pressed: LB (Left Bumper) [310]
Button pressed: RB (Right Bumper) [311]

Button pressed: LT (Left Trigger) [312]
Button pressed: RT (Right Trigger) [313]

Button pressed: A (A) [304] // LG
Button pressed: Y (Y) [308] // RG
```

sc-controller config:
![image](https://user-images.githubusercontent.com/2003017/47444736-d52cab80-d7bf-11e8-87dc-83d92eef932e.png)

---

Someone should test this with other controllers.

**Comment by MaoDzeDun** [comment]
I'll try to do it tonight after work - I've Steam Controller, Xbox 360 Controller, DualShock 4 and some generic Xbox360 gamepad. Will post how it went. I think my 360 was working fine the other day but I'll have to get back to you on that.

**Comment by tkashkin** [comment]
Exit controller shortcut is added in f05d464 along with settings redesign and controller improvements. Use `Guide + B` to exit GameHub.

![image](https://user-images.githubusercontent.com/2003017/58766025-811e4c80-8582-11e9-9c4b-7f6d8a1d8af1.png)

It's now possible to disable individual controllers.

**Comment by neuromancer** [comment]
Oh, neat. I will test this as soon as I get home.

**Comment by Hedronmx** [comment]
This looks amazing!

**Comment by DocMAX** [comment]
Fullscreen mode please. I want to replace Steam as a gamescope fullscreen launcher with controller support, thanks.
A replacement for "steam -gamepadui -steamdeck -steamos" command to be exactly. Steam is too overbloaded for me.
